### ISSUE TITLE
Search autosuggestion: No results for multi-word streets

### ISSUE THREAD

**Issue by scaidermern** [osmandapp/OsmAnd#6233]
The search function returns no results when searching for a street consisting of several words and the second word starts with an umlaut.

Example:

First I select the city in the address tab, here "Bad Berka" (located in Thuringia, Germany). Then I choose to select the street, here "Am Ärzteheim". Typing "Am " even suggests the street I'm looking for:

<table style="width:100%">
  <tr>
<th><img src="https://user-images.githubusercontent.com/1040784/47656883-53a69600-db90-11e8-99d1-44c39f55a8b8.jpg"  width="300" height="auto"/> </th>
  </tr>
</table>  

Typing "Am Ä" fails to find anything at all:

<table style="width:100%">
  <tr>
<th><img src="https://user-images.githubusercontent.com/1040784/47656909-5e612b00-db90-11e8-8298-6326c9babba4.jpg"  width="300" height="auto"/> </th>
  </tr>
</table>  

So far it looks like this only happens when the street name consists of several words and the _second_ word starts with an umlaut. Streets consisting of a single word starting with an umlaut seem to work. Streets consisting of several words with no umlaut at the beginning work, too.

**Comment by sonora** [comment]
It is not limited to the Umlaut, but the behavior changes a little: Search for e.g any Italian town, e.g. Bardolino, and select it, afterwards type Via - you will see dozens of results. Now proceed typing to 'Via a', and you will still see all Vias whose name starts with a. But now if you add more characters, all search results will disappear (now reporting no search results found) until the point where you have finished typing the entire name (e.g. bardolino via adamello ) at which point this result will appeear again).

**Comment by scaidermern** [comment]
You are correct, this also happens for other multi-word street names. But also in large cities. Another example: city Dresden, street Alter Jagdweg. "Dresden Alter J" shows this street in the list of suggestions, ""Dresden Alter J**a**" shows no results at all, "Dresden Alter Jagdweg" finally shows it again.

**Comment by sonora** [comment]
I just tested it also seems to happen in a similar fashion when you tap "Use Online Search", so it seems not a bug in the search engine, but maybe in how we process/provide the search string to the engines, like maybe we add a false blank to the end under some conditions...

**Comment by sonora** [comment]
Behavior is like this: Any word specified after a blank is never matched to an addres field where the word before the blank already produced search results. So if you specify "bad ber", you will not find the town of Bad Berka in the results, but instead only all places which start with "bad" and where another adress field like a streetname exists starting with "ber". Meaning we somehow need to loosen how we treat a blank as a separator in our search string, it must not necessarily be interpreted to indicate the end of an address field.

**Comment by sonora** [comment]
@vshcherb Solution idea: Things may work ok without code changes if on our index we used no-break space characters (or some other character) instead of regular space characters in multi word names. Because I observe that the Bavarian  town of Garmisch-Partenkirchen, spelled with a '-' in the middle,  is correctly found by our search with all of the following search strings:
- garmisch partenkirchen (i.e. with a regular space in the middle)
- garmisch par (space and partial second word, which is the core of this issue)
- parten (i.e. it also works if you just search for the second word)
Only difference it seems between Garmisch-Partenkirchen where things work and e.g. Bad Berka where it fails is the separating character used in the middle, hence my idea.

**Comment by vshcherb** [comment]
I didn't understand the issue itself. Is it about autosuggestion or about search. As I see you didn't enter full street name, so autosuggestion could fail in that situation but it is a smaller issues comparing to search

**Comment by sonora** [comment]
This is not primarily about autosuggestion, but about the search itself. I would think it is one of the **major issues** we have with the Search currently: When you type the initial portion of a (place or street) name correctly, then pause long enough, users expect to see a list of remaining search hits fitting the filter, which would get more refined once they proceed typing. But if the results list is empty already at some point (and even reports that not results can be found), noone expects that typing additional characters would eventually result in new search results, this is counter intuitive, you would just abort the search.

Things work as expected for one-word towns and one word street names. But once either one of them is a compound (e.g. Los Angeles, San Francisco, Bad Berka) things fail. I guess if you scan the list of failed searches people report to us, you will definitely find examples, potentially many?

**Comment by vshcherb** [comment]
I don't see an issue with "Bad Berka" or "Los Angeles". I see an issue with "Bad Berka A" though it is not a search because nothing could be found with such input, it is an autosuggestion issue. Let's separate 2 things to avoid confusion.

**Comment by sonora** [comment]
Yes, there is a reproducibility problem to some degree, I think the issue may manifest a little differently depending on which maps you have installed.

For better reproducibility select the address search "Select city first" option, so we limit this to place names (but it is a universal issue). For me
- typing `bad ber` **does** produce Bad Berka (amongst other hits), but 
- typing `bad berk` **does not** produce "Bad Berka"anymore.
- Proceeding to `bad berka` does again show the result correctly.
Proceeding to streets names:
- `bad berka a` does for me produce correct results, mostly streets in bad berka which start with an 'a', or one of their compound words starts with an 'a'.
- What also works for me is `bad berka am an`, correctly producing "Am Anger, Bad Berka",
- but what fails is `bad berka am ang`, not yielding this result!
- Finally `bad berka am anger` correctly produces the result again.

Very weird, I think this likely demonstrates the major weakness of our current quick search.

**Comment by scaidermern** [comment]
> But if the results list is empty already at some point (and even reports that not results can be found), noone expects that typing additional characters would eventually result in new search results

This is exactly the problem. The user sees a working autosuggestion when starting to type. But then suddenly it stops working and tells the user that there are no results. The user will stop typing at this point.

**Comment by vshcherb** [comment]
> Yes, there is a reproducibility problem to some degree, I think the issue may manifest a little differently depending on which maps you have installed.
That sounds reasonable to algorithm what we have:
"bad berka am anger" is really complex case consisted of many small words.

**Comment by scaidermern** [comment]
With OsmAnd 3.5.5 this bug still persists when using the third tab(address) in the search menu and entering the city, then the street.

Directly searching for "Bad Berka Am Ä" doesn't lead to an empty list of search results. However "Bad Berka Am Ärzteheim" isn't contained in the search results either.

**Comment by Zirochkabila** [comment]
# Testing 1

> OsmAnd Maps 5465, released 16.09.26
> OsmAnd~ 5.5.0#8833m, released: 2026-09-16

| # | Search Engine | Search Query | Expected Result | Actual Result | Status |
| :--- | :--- | :--- | :--- | :--- | :---: |
| **1a** | **Address Search** *(iOS / Android)* | `Bad Berka Am` | Show `Am Ärzteheim` street in suggestion list | Street is present in the suggestion list | 🟢 **PASSED** |
| **1b** | **Address Search** *(iOS / Android)* | `Bad Berka Am Ä` | Keep `Am Ärzteheim` street in the list | Street stays in the list (ranked 3rd) | 🟢 **PASSED** |
| **1c** | **Address Search** *(iOS / Android)* | `Bad Berka Am Ärzteheim` | `Am Ärzteheim` street takes 1st position | Street moves to the 1st position in results | 🟢 **PASSED** |
| **2a** | **Spatial Search** *(Android)* | `Bad Berka Am` | `Am Ärzteheim` present in top POI/street results | Street is missing from top results | 🔴 **FAILED** |
| **2b** | **Spatial Search** *(Android)* | `Bad Berka Am Ä` | Retain `Am Ärzteheim` street in results | Result drops; unrelated POIs/routes are shown | 🔴 **FAILED** |
| **2c** | **Spatial Search** *(Android)* | `Bad Berka Am Ärzteheim` | Street appears in search results | Street appears suddenly after typing full name | 🟢 **PASSED** |




<details><summary>Details</summary>
<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/c60f8b8c-1dee-4fb5-bf8d-5bd67ffec73a"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/3e26a80d-a13f-4023-b461-212a0190b8ee"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/efaa90cb-1f4a-4b31-9a35-e5ccee4bf664"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/fc814811-8664-46d1-acec-cf33a0ff5737"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/b737004b-1ddf-4134-ae93-0010b16fc920"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/bb59b3cd-f708-482e-b233-8534a856aaa2"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/9028ef60-1c36-43f6-b05b-e317231dbfa3"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/4686f8c4-d012-4189-a9b4-8b80f6416e2f"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/3bf3dadf-c30a-474b-ac2a-2a703ac22faf"  width="300" height="auto"/> </th>
  </tr>
</table>  
</details>

**Comment by vshcherb** [comment]
3 letters is enough to be displayed for suggestion - "Bad Berka Am Arz"
